WorkflowSwiftUI/Sources/ObservableScreen.swift

```diff
@@ -0,0 +1,208 @@
+}
+
+private struct ViewEnvironmentModifier: ViewModifier {
+    @ObservedObject var holder: ViewEnvironmentHolder
+
+    func body(content: Content) -> some View {
+        content
+            .environment(\.viewEnvironment, holder.viewEnvironment)
+    }
+}
+
+private final class ViewEnvironmentHolder: ObservableObject {
+    @Published var viewEnvironment: ViewEnvironment
+
+    init(viewEnvironment: ViewEnvironment) {
+        self.viewEnvironment = viewEnvironment
+    }
+}
+
+private final class ModeledHostingController<Model, Content: View>: UIHostingController<ModifiedContent<Content, ViewEnvironmentModifier>> {
+    let setModel: (Model) -> Void
+    let setViewEnvironment: (ViewEnvironment) -> Void
+
+    var swiftUIScreenSizingOptions: SwiftUIScreenSizingOptions {
+        didSet {
+            updateSizingOptionsIfNeeded()
+
+            if !hasLaidOutOnce {
+                setNeedsLayoutBeforeFirstLayoutIfNeeded()
+            }
+        }
+    }
+
+    private var hasLaidOutOnce = false
+
+    init(
+        setModel: @escaping (Model) -> Void,
+        viewEnvironment: ViewEnvironment,
+        rootView: Content,
+        sizingOptions swiftUIScreenSizingOptions: SwiftUIScreenSizingOptions
+    ) {
+        let viewEnvironmentHolder = ViewEnvironmentHolder(viewEnvironment: viewEnvironment)
+
+        self.setModel = setModel
+        self.setViewEnvironment = { viewEnvironmentHolder.viewEnvironment = $0 }
+        self.swiftUIScreenSizingOptions = swiftUIScreenSizingOptions
+
+        super.init(
+            rootView: rootView
+                .modifier(ViewEnvironmentModifier(holder: viewEnvironmentHolder))
+        )
+
+        updateSizingOptionsIfNeeded()
+    }
+
+    @available(*, unavailable)
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("not implemented")
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        // `UIHostingController`'s provides a system background color by default. In order to
+        // support `ObervableModelScreen`s being composed in contexts where it is composed within another
+        // view controller where a transparent background is more desirable, we set the background
+        // to clear to allow this kind of flexibility.
+        view.backgroundColor = .clear
+
+        setNeedsLayoutBeforeFirstLayoutIfNeeded()
+    }
+
+    override func viewDidLayoutSubviews() {
+        super.viewDidLayoutSubviews()
+
+        defer { hasLaidOutOnce = true }
+
+        if #available(iOS 16.0, *) {
+            // Handled in initializer, but set it on first layout to resolve a bug where the PCS is
+            // not updated appropriately after the first layout.
+            // UI-5797
+            if !hasLaidOutOnce,
+               swiftUIScreenSizingOptions.contains(.preferredContentSize)
+            {
+                let size = view.sizeThatFits(view.frame.size)
+
+                if preferredContentSize != size {
```

Comment:
Do we want this in `viewDidLayoutSubviews` or `viewWillLayoutSubviews`? We've rightly run into problems in Market doing layout updates too late, in `viewDidLayoutSubviews`, when `viewWillLayoutSubviews` is more correct.

Reply:
On the other hand, it might be nice to add more "debt" to that ticket. Thoughts?